Regularization

congxie1108 opened the pull request:
*Description of changes:*

Implemented ZoneoutCell (both mxnet version), RNNZoneoutCell (the version fixes the issue that the output to the next layer and in the states using different masks), activation regularization, and temporal activation regularization.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

YuntianYeAWS committed:
SimpleFeedForward to produce DistributionForecast (#870)

* add distribution_network

* add distribution output

* Update estimator.py

* Update _estimator.py

* Update estimator.py

* Update _estimator.py

* Update _estimator.py

* Update _network.py

* Update _estimator.py

* Add files via upload

* Add files via upload

* Update _network.py

* Add files via upload

* Add files via upload

* Add files via upload

* Add files via upload

* Add files via upload

* Update test_model.py

* Update test_model.py

* Update test_model.py

* Update test/model/simple_feedforward/test_model.py

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

* Update src/gluonts/model/simple_feedforward/_network.py

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

* Update _estimator.py

* Update _estimator.py

* Update test/model/simple_feedforward/test_model.py

* Update test/model/simple_feedforward/test_model.py

* Update test/model/simple_feedforward/test_model.py

* Update test/model/simple_feedforward/test_model.py

* Update test/model/simple_feedforward/test_model.py

* Add files via upload

* Add files via upload

* Update test_model.py

* Add files via upload

* Add files via upload

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

Commit:
style check

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

congxie1108 committed:
add activation regularization and temporal activation regularization to deepar

yx1215 committed:
add documentation related to shuffle_buffer_length/ (#910)

fix minor bugs where shuffle_buffer_length is taken as an implicit input

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

congxie1108 committed:
stacking unroll rnncells

congxie1108 committed:
stacking unroll rnncells

congxie1108 committed:
stacking unroll rnncells

congxie1108 committed:
stacking unroll rnncells

congxie1108 committed:
stacking unroll rnncells

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
fix the implementation of zoneout for rnn

congxie1108 committed:
debug

congxie1108 committed:
debug

congxie1108 committed:
preserver output for regularization

congxie1108 committed:
preserver output for regularization

congxie1108 committed:
preserver output for regularization

congxie1108 committed:
preserver output for regularization

lostella committed:
fix multiprocessing in backtest function (#915)

kashif committed:
Scale the negative binomial's gamma (#909)

For issue #906 

The `theta` param. of Gamma gets scale by `scale *theta = alpha * mu * scale`. So only need to scale the `mu` .

benidis committed:
Shape squeeze edge case bug fix (#911)

* edge case with singleton values fix of imputation methods

Commit:
Added support for gzipped files. (#914)

* Added support for gzipped files.

* Fixed gzip.

* Added test for gzip.

congxie1108 committed:
preserver output for regularization

congxie1108 committed:
preserve raw output for regularization

congxie1108 committed:
preserve raw output for regularization

congxie1108 committed:
preserve raw output for regularization

congxie1108 committed:
preserve raw output for regularization

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

congxie1108 committed:
add ar and tar regularization to deepar

dcmaddix committed:
MQCNN: Support for past dynamic features and scaling (#916)

* Adding support for past dynamic features, scaling and default parameters updates for MQCNN

* Fixing black

* Fixing train method signture for shuffle_buffer_length

* Removed overloaded train method

Co-authored-by: Danielle Robinson <dmmaddix@amazon.com>

Commit:
Use of \n to split lines in batch transform. (#920)

`.splitlines()` also splits on `\u2028`, which is not intended since
it can occucr inside a field of the request.

lostella committed:
Fix LDS distribution for the case where sequence_length==1 (#921)

* wrap mx.nd.split to address edge case

* fix import

* update docstring

congxie1108 committed:
clean up

dcmaddix committed:
Fixing cardinality array when use_feat_static_cat = False but feat_static_cat present in dataset (#918)

Co-authored-by: Danielle Robinson <dmmaddix@amazon.com>

Commit:
Defaulting MSIS to NaN when it can't be calculated. (#923)

Commit:
Fix batch-transform case, where request is empty. (#927)

* Fix batch-transform case, where request is empty.

* Update src/gluonts/shell/serve/app.py

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

lostella committed:
making backtest_metrics simpler (#924)

szhengac commented on src/gluonts/model/deepar/_network.py:
```diff
@@ -15,6 +15,7 @@
 
 # Third-party imports
 import mxnet as mx
+import gluonnlp as nlp
```
I don't see where we use gluonnlp in the file

congxie1108 replied:
You are right. My mistake.
I was trying to use AR and TAR from gluonnlp, but ended up with implementing my own version, and forgot to delete the import.

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+from mxnet.gluon import tensor_types
+from mxnet.gluon.loss import Loss
+
+class ZoneoutCell(ModifierCell):
```
Add docstring to each parameter

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+
+        return output, new_states
+
+class RNNZoneoutCell(ModifierCell):
```
Why do we have two zoneout cells?

congxie1108 replied:
Zoneout is functioning the same as the original one in mxnet, which uses different masks for output and the output in states
RNNZoneout uses the same mask for the output and the output inside states (the first element of states)
And for both of them, I add a flag "preserve_raw_output" to indicate whether to append the output without dropout to the states, that's why I also need to implement not only RNNZoneout, but also the original Zoneout.

Commit:
Default QuantileForecast.mean to p50. (#930)

szhengac commented on src/gluonts/model/deepar/_network.py:
```diff
@@ -425,6 +464,27 @@ def hybrid_forward(
+            outputs_raw = []
+            outputs_dropped = []
+            for state, state_info in zip(self.states, self.rnn.state_info()):
+                if 'state_index' in state_info:
```
so the format is [dropped_h, c, h]?

congxie1108 replied:
for each rnn layer, after accumulation, the final version of states will be [dropped_h, c, raw_h, [list of dropped_h], [list of raw_h]]
since there are 2 layers, the states of them will be concatenated.
so there should be totally 10 elements in the final states.

congxie1108 replied:
the logic is:
for raw lstm, the states if [dropped_h, c]
after add zoneout with the option "preserve_raw_output=True", it will append raw_h to the states, then the states become [dropped_h, c, raw_h]
then, after add AccumulateStatesCell with option "index_list=[0, -1]", it will accumulate dropped_h and raw_h throughout timestamps and append them to states, so finally the states become [dropped_h, c, raw_h, [list of dropped_h], [list of raw_h]]

szhengac replied:
10 elements?

congxie1108 replied:
deepar has 2 rnn layers by default
every rnn layer has 5 elements, concatenating 2 layers gets 10.

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+        if self.preserve_raw_output:
+            # the output is stored in the last element of states, thus skip it
+            new_states = [F.where(mask(p_states, new_s), new_s, old_s) for new_s, old_s in
+                                  zip(next_states, states[0:-1])] if p_states != 0. else next_states
```
can remove 0, [:-1] is enough

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+    """
+    def __init__(self, base_cell, index_list):
+        super(AccumulateStatesCell, self).__init__(base_cell)
+        self.index_list = index_list
```
what is index list?

congxie1108 replied:
to indicate which elements in states needs to be accumulated

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+
+    def hybrid_forward(self, F, inputs, states):
+        # identity mapping
+        next_output, next_states = self.base_cell(inputs, states)
```
states here has 3 elements right?

congxie1108 replied:
yes, next_states should only have 3 elements

congxie1108 replied:
states have 5 elements

szhengac replied:
btw, the base_cell accepts number of elements more than the predefined? for example, lstm only has 2 states, but we input 5 elements to it.

congxie1108 replied:
states have 5 elements, and next_states have 3 elements at first, and then have 5 elements when it is returned.
I've confirmed it by adding some assert to debug.

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+            accumulated_states.append(states[len(states)-len(self.index_list)+i] + [next_states[index]])
+        next_states.extend(accumulated_states)
+
+        return next_output, next_states
```
the next_states should be passed to the next time stamp, but it is accumulated?

congxie1108 replied:
yes, when execute "next_output, next_states = self.base_cell(inputs, states)", although there are 5 elements in states, the base cell will ignore the elements on the tail of states, and only use the first 2 elements of states, then return the next_states of [dropped_h, c, raw_h]

szhengac commented on src/gluonts/mx/rnn_extend/rnn_cell.py:
```diff
@@ -0,0 +1,338 @@
+        # append accumulated states
+        accumulated_states = []
+        for i, index in enumerate(self.index_list):
+            accumulated_states.append(states[len(states)-len(self.index_list)+i] + [next_states[index]])
```
it seems that the first part can be reduced to states[i - len(self.index_list)].

congxie1108 committed:
Implemented model iteration averaging to reduce model variance (#901)

* Implemented model iteration averaging to reduce model variance. Two methods are included: NTA and Alpha_Suffix.

lostella committed:
add nan support to simple feedforward model (#933)

timforby committed:
Addded trimming to encoded sagemaker parameters in shell package. (#917)

* Addded trimming to encoded sagemaker parameters.

SageMaker API docs note a hard limit on sagemaker parameters:
https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_HyperParameterTrainingJobSummary.html

We can encode the sagemaker parameters and trim them to a specific length.
Then on decode we reconstruct the parameters.

* Apply suggestions from code review

Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>

* Update src/gluonts/shell/sagemaker/params.py

Co-authored-by: Jasper <jasper.b.schulz@googlemail.com>

* docstring fixes

* Update src/gluonts/shell/sagemaker/params.py

Co-authored-by: Jasper <jasper.b.schulz@googlemail.com>

* more readable docstrings, cleaner trimming

* whitespace fix

* doctest whitespace fix. import count from itertools.

* moving doctest normalize whitespace to setup.cfg

* moving doctest normalize whitespace to pytest

* Update setup.cfg

Co-authored-by: Jasper <jasper.b.schulz@googlemail.com>

Co-authored-by: tim <tjforbes@amazon.com>
Co-authored-by: Jasper <jasper.b.schulz@googlemail.com>
Co-authored-by: Lorenzo Stella <lorenzostella@gmail.com>
Co-authored-by: Jasper <schjaspe@amazon.de>

Schmedu committed:
Shell: Fix writing of `output/failure` file in case of error in provided hyper-parameters. (#942)

* fix for missing environment variables that are not printed to the `failure` file

* refactored to use TrainPaths as parameter

* refactored TrainEnvs usage in tests

* Update src/gluonts/shell/__main__.py

Co-authored-by: Eduard Uffelmann <uffeledu@amazon.com>
Co-authored-by: Jasper <jasper.b.schulz@googlemail.com>

congxie1108 committed:
test different droput strategies
